### PR TITLE
feat: migrate JsonIgnore to JsonView

### PR DESCRIPTION
**Pull request by sidhantgoel**
In summary, the change comprises adding `JsonView` annotations on domain/model fields, controller methods, and arguments.
applying JsonView annotation on controller methods will include fields which are marked with the same view in domain objects.
Similarly, usage on controller method arguments will only deserialize the fields which are marked with same view in Domain object, which is helpful in case we don't want the client to send some fields for security reasons.
This change replaces the previous use of JsonIgnore for restricting fields in API response and allows for more flexibility and will enable us to have fine-grain control over fields serialized for different contexts, such as API response, export, and import.
The following views are defined as of now.
`View.Internal` - View used to serialize for internal use. It inherits from `Views.Public` and `Views.Export`, so it will also include those fields.
`Views.Public` - View used for API request/response.
`Views.Export` - View used for making field exportable, like in the case of Git sync and file export.
Also, we can define more views by adding an interface inside the Views class and Views can also inherit fields marked with different view by simple interface inheritance.

A small tutorial on JsonView - [Jackson JSON Views | Baeldung](https://www.baeldung.com/jackson-json-view-annotation)

### CODE REVIEW

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/Views.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    /**
+     * Used with JsonView annotation to denote that a field should be visible to the client.
+     */
+    interface Public {}
+
+    /**
+     * Used with JsonView annotation to denote that a field should be exported when exporting the Domain/DTOs.
+     */
+    interface Export {}
+}
```
Comment (abhvsn) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/Views.java:21]: Does it make sense to have a separate `view` package under `com.appsmith.server` and not mix with models?

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/dtos/ExecuteActionDTO.java [hunk 2]
```diff
@@ -15,41 +18,50 @@
...
+     * "k4": {
+     * "array": [
+     * "string",
+     * "number",
+     * "string",
+     * "boolean"
+     * ]
+     * },
+     * "k5": "boolean"
+     * }
```
Comment (abhvsn) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/dtos/ExecuteActionDTO.java:48]: With earlier indentation, it's more easy to understand the structure.

**Review by abhvsn — COMMENTED** [review]
@sidhantgoel QQ: With JsonView we should not require separate models for export. If this assumption is correct why are we maintaining these models as well.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/dtos/ExecuteActionDTO.java (hunk 2 above)
Comment (sidhantgoel) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/dtos/ExecuteActionDTO.java:48]: Auto format did this, will revert

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/Views.java (hunk 1 above)
Comment (sidhantgoel) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/Views.java:21]: > @sidhantgoel QQ: With JsonView we should not require separate models for export. If this assumption is correct why are we maintaining these models as well.

Which model you are talking about?

Comment (sidhantgoel) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/Views.java:21]: We can have a separate package but that will contain only single class, I don't think it is a problem though, will move this

Comment (abhvsn) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/models/Views.java:21]: > > @sidhantgoel QQ: With JsonView we should not require separate models for export. If this assumption is correct why are we maintaining these models as well.
> 
> Which model you are talking about?

@sidhantgoel I was referring to models under `app/server/appsmith-server/src/main/java/com/appsmith/server/models/export` package.

**Comment by abhvsn** [comment]
LGTM 
@sidhantgoel can you please resolve the merge conflicts before I approve?

**Review by nayan-rafiq — APPROVED** [review]

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/ActionCollectionServiceImplTest.java [hunk 3]
```diff
@@ -453,10 +454,10 @@ public void testUpdateUnpublishedActionCollection_withInvalidId_throwsError() th
     @Test
     public void testUpdateUnpublishedActionCollection_withModifiedCollection_returnsValidCollection() throws IOException {
         ObjectMapper objectMapper = new ObjectMapper();
-        final JsonNode jsonNode = objectMapper.readValue(mockObjects, JsonNode.class);
-        final ActionCollection actionCollection = objectMapper.convertValue(jsonNode.get("actionCollectionWithAction"), ActionCollection.class);
-        final ActionCollectionDTO modifiedActionCollectionDTO = objectMapper.convertValue(jsonNode.get("actionCollectionDTOWithModifiedActions"), ActionCollectionDTO.class);
-        final ActionCollection modifiedActionCollection = objectMapper.convertValue(jsonNode.get("actionCollectionAfterModifiedActions"), ActionCollection.class);
+        final JsonNode jsonNode = objectMapper.readerWithView(Views.Public.class).readValue(mockObjects, JsonNode.class);
+        final ActionCollection actionCollection = objectMapper.readerWithView(Views.Public.class).readValue(objectMapper.writerWithView(Views.Public.class).writeValueAsString(jsonNode.get("actionCollectionWithAction")), ActionCollection.class);
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/test/java/com/appsmith/server/services/ActionCollectionServiceImplTest.java:458]: Nit, could you please create a utility function for this? I think we'll end up writing this kind of representation for a lot of our tests moving forward.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/models/export/ActionConfigurationMetadata.java [hunk 4]
```diff
@@ -1,64 +0,0 @@
-package com.appsmith.server.models.export;
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/models/export/ActionConfigurationMetadata.java:1]: QQ for my understanding, why are we removing all these classes?

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/views/Views.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+    /**
+     * Used with JsonView annotation to denote that a field should be visible to the client.
+     */
+    interface Public {}
+
+    /**
+     * Used with JsonView annotation to denote that a field should be exported when exporting the Domain/DTOs.
+     */
+    interface Export {}
```
Comment (nidhi-nair) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/views/Views.java:20]: I think you mentioned this before, but will consumption of this view be part of another PR?

**Comment by Parthvi12** [comment]
Bugs filed during sanity testing:

**Comment by Parthvi12** [comment]
Bugs found while sanitty testing:


- [ ] https://github.com/appsmithorg/appsmith/issues/21309


- [ ] On running actions such as JSObject/query gives internal server error

Screenshot 2023-03-02 at 2 16 45 PM

**Comment by Parthvi12** [comment]
https://github.com/appsmithorg/appsmith/issues/21309 is observed on prod as well, hence not a blocker for this PR

**Review by nayan-rafiq — APPROVED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/models/export/ActionConfigurationMetadata.java (hunk 4 above)
Comment (sidhantgoel) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/models/export/ActionConfigurationMetadata.java:1]: These classes were used temporarily, now since we have multiple views, existing classes can be used.

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/views/Views.java (hunk 5 above)
Comment (sidhantgoel) [reply app/server/appsmith-interfaces/src/main/java/com/appsmith/external/views/Views.java:20]: Yes

File: app/server/appsmith-server/src/test/java/com/appsmith/server/services/ActionCollectionServiceImplTest.java (hunk 3 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/test/java/com/appsmith/server/services/ActionCollectionServiceImplTest.java]: @nidhi-nair I've created a utility class to wrap the reading and writing methods of object mapper. Would you please review this once again?

**Review by nidhi-nair — APPROVED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ObjectMapperUtils.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    public <T> T readFromFile(File file, Class viewsClass, Class<T> resultClass) throws IOException {
+        return objectMapper
+                .readerWithView(viewsClass)
+                .readValue(file, resultClass);
+    }
+
+    public String writeAsString(Object src, Class viewsClass) throws JsonProcessingException {
+        return objectMapper
+                .writerWithView(Views.Public.class)
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ObjectMapperUtils.java:31]: Oops looks like this method is not using the parameter?

Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ObjectMapperUtils.java]: Sorry, I missed that. Thanks a lot for your review.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ObjectMapperUtils.java [hunk 7]
```diff
@@ -6,7 +6,12 @@
 import java.io.File;
 import java.io.IOException;
 
-public record ObjectMapperUtils(ObjectMapper objectMapper) {
+public class ObjectMapperUtils {
+    private final ObjectMapper objectMapper;
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ObjectMapperUtils.java:10]: Should we autowire this from the default? There's a few settings there that might come in handy. Includes handling some types like `Instant` and such.

Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/ObjectMapperUtils.java:10]: @nidhi-nair we can do that in future if required. As of now, we don't need to do that because it's only used in tests where an external objectmapper is used.

**Review by nidhi-nair — APPROVED** [review]

**Review by nidhi-nair — APPROVED** [review]
